pallets/msa/src/offchain_storage.rs

```diff
@@ -0,0 +1,464 @@
+		let default_starting_block_number = block_number
+			.saturating_sub(BlockNumberFor::<T>::from(NUMBER_OF_PREVIOUS_BLOCKS_TO_CHECK));
+		let mut start_block_number = last_processed_block_storage
+			.get::<BlockNumberFor<T>>()
+			.unwrap_or(Some(default_starting_block_number))
+			.unwrap_or(default_starting_block_number);
+
+		// since this is the last processed block number we already processed it and starting from the next one
+		start_block_number += BlockNumberFor::<T>::one();
+		while start_block_number <= block_number {
+			if reverse_map_msa_keys::<T>(start_block_number) {
+				if guard.extend_lock().is_err() {
+					log::warn!("last processed block lock is expired in block {:?}", block_number);
+					break
+				}
+			}
+			last_processed_block_storage.set(&start_block_number);
+			start_block_number += BlockNumberFor::<T>::one();
+		}
+	} else {
+		log::info!("skipping processing events on {:?} due to existing lock!", block_number);
+	};
+}
+
+/// Set offchain index value, used to store MSA Events to be process by offchain worker
+fn set_offchain_index<V>(key: &[u8], value: V)
+where
+	V: Encode + Clone + Decode + Eq + Debug,
+{
+	offchain_index::set(key, value.encode().as_slice());
+}
+
+/// Get offchain index value, used to store MSA Events to be process by offchain worker
+fn get_offchain_index<V>(key: &[u8]) -> Option<V>
+where
+	V: Encode + Clone + Decode + Eq + Debug,
+{
+	let value = offchain_common::get_index_value::<V>(key);
+	value.unwrap_or_else(|e| {
+		log::error!("Error getting offchain index value: {:?}", e);
+		None
+	})
+}
+
+/// Offchain indexed compatible Event type
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebugNoBound)]
+enum IndexedEvent<T: Config> {
+	/// A new Message Service Account was created with a new MessageSourceId
+	IndexedMsaCreated {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+
+		/// The key added to the MSA
+		key: T::AccountId,
+	},
+	/// An AccountId has been associated with a MessageSourceId
+	IndexedPublicKeyAdded {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+
+		/// The key added to the MSA
+		key: T::AccountId,
+	},
+	/// An AccountId had all permissions revoked from its MessageSourceId
+	IndexedPublicKeyDeleted {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+		/// The key no longer approved for the associated MSA
+		key: T::AccountId,
+	},
+}
+
+impl<T: Config> IndexedEvent<T> {
+	/// maps a pallet event to indexed event type
+	pub fn map(event: &Event<T>, msa_id: MessageSourceId) -> Option<Self> {
+		match event {
+			Event::MsaCreated { msa_id, key } =>
+				Some(Self::IndexedMsaCreated { msa_id: *msa_id, key: key.clone() }),
+			Event::PublicKeyAdded { msa_id, key } =>
+				Some(Self::IndexedPublicKeyAdded { msa_id: *msa_id, key: key.clone() }),
+			Event::PublicKeyDeleted { key } =>
+				Some(Self::IndexedPublicKeyDeleted { msa_id, key: key.clone() }),
+			_ => None,
+		}
+	}
```

Comment:
Should we check that the `msa_id` passed in matches the `msa_id` in the event in the case of `Event::MsaCreated` and `Event::PublicKeyAdded`?

Reply:
it doesn't matter tbh. We are always using the msa provided by event (you can tell by `*msa`) if it exists and only use the param one for delete which don't have any msa_id.

Reply:
Ah, I see... nit: maybe rename the function input parameter to be distinct from `msa_id` inside the event? It caused me some confusion.